### PR TITLE
Add support for azure

### PR DESCRIPTION
**Pull request by nickjiang2378**
I accidentally committed some large files, so you won't see the commit history because I copied over the changes onto a fresh clone.

### CODE REVIEW

**Comment by MingweiSamuel** [comment]
For commit names use https://www.conventionalcommits.org/en/v1.0.0/ and for lints run `cargo +nightly fmt` and `cargo clippy --all-targets`

**Comment by MingweiSamuel** [comment]
For editing and rearrange old commits look into `git rebase -i origin/main`

File: hydro_deploy/core/src/azure.rs [hunk 1]
```diff
@@ -0,0 +1,570 @@
...
+use super::ssh::LaunchedSSHHost;
+use super::terraform::{TerraformOutput, TerraformProvider, TERRAFORM_ALPHABET};
+use super::util::async_retry;
+use super::{
+    ClientStrategy, Host, HostTargetType, LaunchedHost, ResourceBatch, ResourceResult,
+    ServerStrategy,
+};
+
+// #[derive(Debug)]
+pub struct LaunchedComputeEngine {
```
Comment (shadaj) [inline hydro_deploy/core/src/azure.rs:25]: Should probably be renamed to whatever Azure calls this (Compute Engine is a GCP term?). Also commented line above should be removed.

File: hydro_deploy/core/src/azure.rs [hunk 2]
```diff
@@ -0,0 +1,570 @@
...
+            .unwrap()
+            .path()
+            .join(".ssh")
+            .join("vm_instance_ssh_key_pem")
+    }
+}
+
+#[async_trait]
+impl LaunchedSSHHost for LaunchedComputeEngine {
+    fn server_config(&self, bind_type: &ServerStrategy) -> ServerBindConfig {
```
Comment (shadaj) [inline hydro_deploy/core/src/azure.rs:47]: This is _nearly_ identical to GCP, so I wonder if we should move this into the trait and define a separate abstract function for getting the `internal_ip`.

File: hydro_deploy/core/src/azure.rs [hunk 3]
```diff
@@ -0,0 +1,570 @@
...
+
+    fn resource_result(&self) -> &Arc<ResourceResult> {
+        &self.resource_result
+    }
+
+    fn ssh_user(&self) -> &str {
+        self.user.as_str()
+    }
+
+    async fn open_ssh_session(&self) -> Result<AsyncSession<TcpStream>> {
```
Comment (shadaj) [inline hydro_deploy/core/src/azure.rs:84]: Same here, this is identical to GCP so we should share the logic.

File: hydro_deploy/hydro_cli_examples/hydro_python_sender_azure.hydro.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    # gcp_vpc = hydro.GCPNetwork(
+    #     project="autocompartmentalization",
+    # )
+
+    machine2 = deployment.AzureHost(
+        project="hydro-example" + str(int(time.time())),
+        os_type="linux",
+        machine_size="Standard_B1s",
+        region="East US",
+        image="",
```
Comment (shadaj) [inline hydro_deploy/hydro_cli_examples/hydro_python_sender_azure.hydro.py:20]: If the image can indeed be omitted, we should make this an optional parameter.

File: hydro_deploy/hydro_cli_examples/hydro_python_sender_azure.hydro.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+from pathlib import Path
+import time
+
+async def main(args):
+    machine_azure = args[0] == "azure"
+
+    deployment = hydro.Deployment()
+    localhost_machine = deployment.Localhost()
+
+    # gcp_vpc = hydro.GCPNetwork(
```
Comment (shadaj) [inline hydro_deploy/hydro_cli_examples/hydro_python_sender_azure.hydro.py:11]: Leftover code

**Review by shadaj — CHANGES_REQUESTED** [review]

File: hydro_deploy/hydro_cli_examples/hydro_python_sender_azure.hydro.py (hunk 4 above)
Comment (nickjiang2378) [reply hydro_deploy/hydro_cli_examples/hydro_python_sender_azure.hydro.py]: The azure image parameter is a little lengthy, so I made it optional for the user.

https://arc.net/l/quote/moucnywd

File: hydro_deploy/core/src/azure.rs (hunk 3 above)
Comment (nickjiang2378) [reply hydro_deploy/core/src/azure.rs]: I moved the launchedComputeEngine into a shared file

File: hydro_deploy/core/src/azure.rs [hunk 6]
```diff
@@ -0,0 +1,486 @@
...
+    user: String,
+    pub internal_ip: String,
+    pub external_ip: Option<String>,
+}
+
+impl LaunchedSSHHost for LaunchedVirtualMachine {
+    fn get_external_ip(&self) -> Option<String> {
+        self.external_ip.clone()
+    }
+    fn get_internal_ip(&self) -> String {
```
Comment (shadaj) [inline hydro_deploy/core/src/azure.rs:27]: nit: newlines between functions

File: hydro_deploy/core/src/gcp.rs [hunk 7]
```diff
@@ -1,141 +1,42 @@
...
-            ServerStrategy::Tagged(underlying, id) => ServerBindConfig::Tagged(
-                Box::new(LaunchedSSHHost::server_config(self, underlying)),
-                *id,
-            ),
-            ServerStrategy::Null => ServerBindConfig::Null,
-        }
+    fn get_external_ip(&self) -> Option<String> {
+        self.external_ip.clone()
+    }
+    fn get_internal_ip(&self) -> String {
```
Comment (shadaj) [inline hydro_deploy/core/src/gcp.rs:28]: nit: newlines between functions

**Review by shadaj — CHANGES_REQUESTED** [review]
Just two tiny nits, otherwise looks great!

File: hydro_deploy/core/src/gcp.rs (hunk 7 above)
Comment (nickjiang2378) [reply hydro_deploy/core/src/gcp.rs:29]: done

**Review by shadaj — APPROVED** [review]
Officially multi cloud! 💪

**Pull request merged**